### DISCUSSION TITLE
Dual TPU M.2 Coral Accelerator

### DISCUSSION THREAD

**Discussion by jayburkard** [blakeblackshear/frigate#988] [Ask A Question]
Can't seem to find the USB edge accelerator in stock anywhere. Saw some unraid guys had success using the mini pcie module. Anything conceptually wrong with using the new dual TPU unit with an M.2 -> pcie adapter and passing it all through?

**Comment by hudsonsecurity** [comment]
Doesn't work with most setups to get dual tpu. Stock with the single. Read the github issue open for dual tpu on Google corals page. My opinion you're better off getting a dedicated device with tpu integration.

**Reply by jayburkard** [reply]
Thanks!! Got the single, put it in a mini pci->pci converter on my HP G8 ml350p, hoping to get frigate up in the next week or two

**Reply by digiblur** [reply]
Found the same as well as there just isn't much out there that supports the dual lanes on that port.  It would be cool if they made a full PCIE one with dual on it.  I know they have that 8 and 16 card version but it's a littttttle out of my range too much for my needs.  

You can always roll with two separate boards though.

**Reply by didi767** [reply]
@digiblur did you eventually get it to work?
I have the dual inserted to a pcie card and then to my ESXi as pass through to home assistant VM. But unfortunately it's not being detected. Really frustrating :/

**Reply by jayburkard** [reply]
> @digiblur did you eventually get it to work?
> I have the dual inserted to a pcie card and then to my ESXi as pass through to home assistant VM. But unfortunately it's not being detected. Really frustrating :/

Not dual,  but I had major issues passing through with ESXi. I switched to proxmox and got it to work eventually. 

I have single in pcie card, passing through to ubuntu VM running home assistant

**Reply by didi767** [reply]
So you think that even if I switch to Ubuntu instead of the dedicated vmdk I would still be having issues with the pass through?
I have like 5 vms currently on ESXi and just thinking about transition to proxmox gives me a headache :(
But then again I really want to utilize this dual TPU for frigate

**Reply by digiblur** [reply]
I don't have the dual anymore, abandoned it as I haven't needed to jump to two yet.

**Reply by didi767** [reply]
Yeah, I think I jumped too high as well.
So there is no way I can pass it through on ESXi with Ubuntu 20 server?
Have to go with the proxmox for that?

**Reply by felipecrs** [reply]
I bought the [dual TPU](https://coral.ai/products/m2-accelerator-dual-edgetpu), it boots correctly and if I set like:

```
detectors:
  coral1:
    type: edgetpu
    device: pci:0
```

Frigate seems to use it too:

![image](https://user-images.githubusercontent.com/29582865/144618332-4ac07897-1824-4eb2-8cfd-2c2ba61f57b3.png)

But I have no clue if the two TPUs are being used, or how I would configure Frigate to use both. Any clue on how to configure Frigate to make use of both?

I have a BlitzWolf BW-MPC1 by the way. Don't buy one, because NVME SSDs are not supported and the fan is too noisy, but it at least seems to work with the Coral M.2 card (I replaced the Wi-Fi card with it).

**Reply by T89L** [reply]
It's only using one TPU in the screenshot you are showing there. For the second TPU it would show up as another independent device like "pci:1" and you would add that entry to frigate as well.

**Comment by markmghali** [comment]
Following this as I just bought the dual tpu as it is the only thing in stock. I am on unraid, hoping there is a soltuon?

**Comment by rlrl** [comment]
bought the usb & dual m.2 from arrow about a month ago. their website says 359 parts available for usb (was over 1300 when i ordered)

https://www.arrow.com/en/products/g950-06809-01/google-corporation

**Comment by richardkuw** [comment]
I have the m.2 dual TPU working through Proxmox (since that was the only TPU actually available). I spent a couple of days trying to get pcie passthrough working on Frigate with esxi 7.0. There are issues with MSI-X interrupts that I couldn't resolve so I migrated my server and all my VMs to Proxmox (after running esxi for 12 years!). Proxmox is working fine passing the dual TPU through to a KVM virtual machine. I'm using a Fenvi wifi adapter from Amazon - note there is more than one, the one that works is the 'M.2 NGFF A+E key to PCI Express adapter'. Of course only one TPU is recognized.

I've seen some suggestions to run Frigate on an LXC container in Proxmox but in order to run Docker (a container inside a container) you need to disable ALL of the security features. Docker runs as root. You need to run the container in privileged mode with apparmor disabled. So Docker root is the same as Proxmox root. It is a hideously insecure setup. I tried it briefly but went with a KVM virtual machine and followed the instructions for passthrough. Performance is OK, inference speed is around 9.

I tend to document things so if anyone wants my notes on migrating from esxi to Proxmox and setting up passthrough let me know.

**Reply by tmjpugh** [reply]
sadly it is hard to find m.2 e-key to pcie x2channel. everything is x1channel so only the single tpu is recognized.

**Reply by TheStigh** [reply]
@richardkuw How did you set up the detector setup for the dual TPU ?

**Reply by richardkuw** [reply]
As I mentioned, I'm using a standard wifi adapter from Amazon, so currently only one TPU is recognized. I'm hoping to get one of the new adapters when they are available. That said, if anyone needs instructions to get an m.2 Coral working in Proxmox let me know, I documented it pretty well when I set it up.

**Reply by richardkuw** [reply]
I'll just paste in my Proxmox configuration in case it's useful to anyone. I used a KVM VM called VM 101, change to match your VM. 

Edit /etc/modules on Proxmox host:
add: 
vfio 
vfio_iommu_type1 
vfio_pci vfio_virqfd

Edit /etc/modprobe.d/pve-blacklist.conf on Proxmox host:
add: 
blacklist apex

Update initramfs and reboot:
#update-initramfs -u -k all 
#reboot

Check result:
#lspci -nnk (should show no apex driver in use for TPU on the Proxmox host)

Turn on passthrough for VM 101 using pcie address of TPU. Get that from lspci:
#qm set 101 -hostpci0 01:00.0,pcie=on

Create file to add intel_iommu=on boot parameter (or amd_iommu=on if appropriate). VM WILL NOT BOOT with passthrough without IOMMU. 
Created: /etc/default/grub.d/iommu.cfg:
GRUB_CMDLINE_LINUX="$GRUB_CMDLINE_LINUX intel_iommu=on"

Apply changes: 
#proxmox-boot-tool refresh
#reboot

Results:
#lspci -nnk (on Proxmox host) shows no apex driver in use:
01:00.0 System peripheral [0880]: Global Unichip Corp. Coral Edge TPU [1ac1:089a] Subsystem: Global Unichip Corp. Coral Edge TPU [1ac1:089a] Kernel driver in use: vfio-pci Kernel modules: apex

#lspci -nnk (in VM) shows apex driver being used:
01:00.0 System peripheral [0880]: Global Unichip Corp. Coral Edge TPU [1ac1:089a] Subsystem: Global Unichip Corp. Coral Edge TPU [1ac1:089a] Kernel driver in use: apex Kernel modules: apex

**Reply by T89L** [reply]
> I'm hoping to get one of the new adapters when they are available

Are you aware of any upcoming adapters? I have a dual TPU on the way as well as it's the only thing in stock.

**Reply by richardkuw** [reply]
There is a project to produce adapters for the various Corals in small volumes. It has a wait list, I'm hoping to get one early next year. 
https://github.com/magic-blue-smoke/Dual-Edge-TPU-Adapter

Until then the single TPU is working just fine, my inference speed is around 9 ms (running through Proxmox, which also works just fine). 9 ms should support over 100 fps so with four cameras detecting at 5 fps I doubt I'd actually notice any difference from the second TPU - beyond the warm fuzzies / bragging rights from having a lower inference number.

**Reply by kspearrin** [reply]
@richardkuw your documentation would be appreciated. I see some of it was posted above. Is that all? Also, you keep mentioning using a wifi adapter. I am a bit confused how this relates to the Coral device?

**Reply by richardkuw** [reply]
The Coral m.2 dual TPU is an E-Key m.2 device (the key is just a gap between certain pins in the connector and the letter specifies where the gap is). Typically E-key is used for wifi card adapters, M-key is more commonly used for storage. So the correct adapter to use this particular device in a pcie slot is a wifi card adapter. I used the "fenvi Desktop Wireless Network M.2(NGFF) WiFi Card to PCIe 1X Adapter Converter" from Amazon. I have to warn that the one I bought didn't include the tiny screw to secure the m.2 card to the adapter. Luckily I had one on hand. That gives you access to one of the TPUs, there is no standard adapter yet for both TPUs. 

The documentation I pasted in had to do with the specifics of getting it working with a Proxmox VM, there's lots of documentation for getting the device working with Frigate.

**Reply by kspearrin** [reply]
@richardkuw Ah, makes sense now. My mobo has a M2 E-key slot already, so I don't think I will need an adapter. Maybe both TPUs will work in my case? I am going to attempt to get this working on a Proxmox host as well, so if you have any additional docs or guidance that would be really helpful.

**Reply by T89L** [reply]
@kspearrin There are a bunch of different coral devices. Basically the only one you can find in stock right now (and barely so) is the E-Key dual TPU version. E-Key is usually used for little wifi cards in laptops and to use it in a desktop you most likely need an adapter. Unfortunately these wifi cards don't usually need the hardware the coral dual tpu requires to make both TPUs available, so using these adapters limits you to a single TPU being recognized. The link @richardkuw provided is somebody making custom boards that would allow both TPUs to be recognized.

For now, I am getting by with 1 TPU using a PEXM2150E adapter. Pretty sure the USB coral is just a single TPU anyways so it's basically the same as having one of those.

**Reply by richardkuw** [reply]
Once you follow the steps I outlined you should be able to see the device on the pcie bus in the VM, and then it's like any other install.

**Reply by kspearrin** [reply]
Thanks. Info for others: The only place I could find on the internet with stock of any coral device was this online store, based in the UK: https://export.rsdelivers.com/product/coral/g650-06076-01/coral-m2-accelerator-with-dual-edge-tpu-8-bit-g650/2133255

It was $10 shipping to the US. Hopefully it will arrive soon.

**Reply by T89L** [reply]
@kspearrin That's where I got mine. I think it took about 2 weeks.

**Reply by T89L** [reply]
I finally received my dual TPU adapter and I'm happy to report it's working great! It took over a month to get here from Russia, and shipped just before the war started. Not sure what impact the war will have on shipments now but it appears the seller is working on getting more distribution channels up.

![image](https://user-images.githubusercontent.com/11653955/160003464-9104d1a1-3d53-4cd8-b13e-9b824470ef26.png)

**Reply by felipecrs** [reply]
How do you know that it's indeed using both cores? Because AFAIK you can set as many as you want in Frigate config, right?

How did you set detectors? Like this?

```yaml
detectors:
  coral1:
    type: edgetpu
    device: pci:0
  coral2:
    type: edgetpu
    device: pci:0
```

**Reply by digiblur** [reply]
It is two separate corals.

**Reply by felipecrs** [reply]
Meaning this?

```yaml
detectors:
  coral1:
    type: edgetpu
    device: pci:0
  coral2:
    type: edgetpu
    device: pci:1
```

**Reply by felipecrs** [reply]
```console 
➜ lspci | grep Edge
01:00.0 System peripheral: Global Unichip Corp. Coral Edge TPU
```

Does this mean my motherboard only recognized one TPU from my Dual TPU?

**Reply by didi767** [reply]
> I finally received my dual TPU adapter and I'm happy to report it's working great! It took over a month to get here from Russia, and shipped just before the war started. Not sure what impact the war will have on shipments now but it appears the seller is working on getting more distribution channels up.
> 
> ![image](https://user-images.githubusercontent.com/11653955/160003464-9104d1a1-3d53-4cd8-b13e-9b824470ef26.png)

@tim-lukacik that's incredible! Could you please where did you order this from?

Thanks!

**Reply by felipecrs** [reply]
Thanks a lot, @tim-lukacik. That's the easiest way to confirm that my system only recognizes one of the two TPUs.

By the way, HassOS 8.0 mentions:

> Google Coral support is now using Google’s latest driver. This enables additional Coral device support such as PCI Dual Edge TPU.

https://www.home-assistant.io/blog/2022/05/16/home-assistant-os-release-8/#other-changes

**Reply by didi767** [reply]
@felipecrs but I think we still have to have a 2 lane m.2 or this adapter that splits the pci into two lanes in order to have both TPU, regardless of the driver update.

**Comment by Yonny24** [comment]
Also looking at alternatives for the usb coral.

Is it easier to set up with the standard M.2 accelerator and/or single mini pcie accelerator?
Do they have the same performance as the coral usb?

I have no issues wiping proxmox and installing generic linux and frigate. 
Looking at some mini pc's with m.2 expansion slots or mini pcie slots.

Thanks

**Reply by richardkuw** [reply]
I suggest you find a supplier first. I went with the dual m.2 TPU because it was actually available. Now I don't see stock of any of them (at Mouser). None are expected until March.

I don't think m.2 vs. mini pcie should make any difference as long as you have the appropriate adapter for your motherboard.

**Reply by tmjpugh** [reply]
> Is it easier to set up with the standard M.2 accelerator and/or single mini pcie accelerator? Do they have the same performance as the coral usb?
> 
 I’m my experience PCIe was slightly faster

My server not have USB C so I used PCIe USB port card to add support. I lost connection with usb coral at times. PCIe coral was more stable in this regard as well

single and dual coral setup was same. For me only single shows up with dual coral as well. Hoping to find working adapter in future

**Comment by rchiileea** [comment]
I could not get hte dual tpu to work, when using the config setup like this 

detectors:
  coral1:
    type: edgetpu
    device: pci:0
  coral2:
    type: edgetpu
    device: pci:1

frigate would not load properly and terminate so gave up and ran with just the 1, then I thought I would try this

detectors:
  coral1:
    type: edgetpu
    device: pci:1
  coral2:
    type: edgetpu
    device: pci:0

and it worked lol, go figure

[2023-01-28 18:10:42] frigate.edgetpu                INFO    : Attempting to load TPU as pci:0
[2023-01-28 18:10:42] ws4py                          INFO    : Using epoll
[2023-01-28 18:10:42] frigate.edgetpu                INFO    : Attempting to load TPU as pci:1
[2023-01-28 18:10:42] frigate.app                    INFO    : Output process started: 236
[2023-01-28 18:10:42] frigate.edgetpu                INFO    : TPU found
[2023-01-28 18:10:42] frigate.edgetpu                INFO    : TPU found

**Reply by tmjpugh** [reply]
How are you connecting TPU?  What adapter do you use or no adapter required for you?

**Reply by didi767** [reply]
> I could not get hte dual tpu to work, when using the config setup like this
> 
> detectors: coral1: type: edgetpu device: pci:0 coral2: type: edgetpu device: pci:1
> 
> frigate would not load properly and terminate so gave up and ran with just the 1, then I thought I would try this
> 
> detectors: coral1: type: edgetpu device: pci:1 coral2: type: edgetpu device: pci:0
> 
> and it worked lol, go figure
> 
> [2023-01-28 18:10:42] frigate.edgetpu INFO : Attempting to load TPU as pci:0 [2023-01-28 18:10:42] ws4py INFO : Using epoll [2023-01-28 18:10:42] frigate.edgetpu INFO : Attempting to load TPU as pci:1 [2023-01-28 18:10:42] frigate.app INFO : Output process started: 236 [2023-01-28 18:10:42] frigate.edgetpu INFO : TPU found [2023-01-28 18:10:42] frigate.edgetpu INFO : TPU found

SERIOUSLY????? the order made the difference??? wth

**Reply by rchiileea** [reply]
[this one](https://www.makerfabs.com/dual-edge-tpu-adapter-m2-2280-b-m-key.html)

**Reply by Genieplumb** [reply]
I have this exact card and a dual tpu coral, when running it flat out it will get hot and overheat unless fan cooled with an additional heatsink. Ive got the PCI card also on the way so I can put a fan on it.. just thought I would share my findings.. also ive looked it up and others have had a similar experience with overheating when adding the other tpu

**Reply by Genieplumb** [reply]
this is the card I have on order (has a fan mount) https://www.makerfabs.com/dual-edge-tpu-adapter.html

**Reply by NickM-27** [reply]
If you use it in a spot where the motherboard has a heat sink then it shouldn't be an issue. I have one of each of these (1 PCIe with heatsink / fan and 1 m.2 with motherboard m.2 heatsink on it) and both are roughly the same temp

![Screen Shot 2023-01-29 at 13 01 03 PM](https://user-images.githubusercontent.com/14866235/215352661-c3d83c83-8139-4902-b5c5-e41339d3b36b.png)

**Reply by digiblur** [reply]
Haven't had issues with the dual overheating either but I do recall documents from Google saying that some cooling could be required.

**Reply by rchiileea** [reply]
I actuall just bought a set of 20mm heatsinks with some 2mm thick thermal adheasive tape and stuck than on the top of the whol device its fine.

**Reply by FrigateEnthusiast** [reply]
May sound "hacky" but I've have put small 15mm axial fans (like used on 3D printers extruders) inside cases, directly blowing on components that needed extra cooling help such as a  coral pice devs, and were not in high air flow areas. Easy to power from a MB fan connector or tap into a the 5v or 12v power from a sata drive connector from the PS.

**Comment by Saboral** [comment]
Got my dual TPU working used this adapter:
https://www.makerfabs.com/dual-edge-tpu-adapter.html

Had to blacklist the apex and gasket drivers in modprobe.d on Proxmox, update initramf, reboot, and manually add the PCI passthru to the VM using the shell command. Once that was done I had zero issue passing through to my VM.

Once I went in the VM and added the TPU Runtime and Pycoral I was in business. Currently getting a 11.09 inference speed, but I'm still tweaking my config.

I have a USB Coral as well, but I found it was having a lot of connectivity issues with Proxmox passthru. I'm convinced the USB on the Dell r720 I use wasn't supplying sufficient power for the TPU under load.

**Comment by jlarnal** [comment]
At this point in time (Q4 2023), I end up wondering if the Google Coral team ever had a brainstorming session about the M.2 E-key choice for the dual TPU module. There's so few end products (other than the Frigate NVR) that offer enough (populated) lanes on their M.2 expansion slots !
And all the industrial SFF motherboards that do (have those lanes), cost a leg. Plus, at their price range, why would an engineer bother with int8 inferences anyway ? They'd go with FPU-able NPUs, as most of these board come with NPUs integrated with the onboard SOC. 

Really, the Coral M.2 dual edge TPU module is a head-scratcher to me. No compatibility whatsoever with the run-of-the-mill iot sbcs the Coral is aimed to supplement ( in terms of power figures ). 🤔

I'm weeping over the fact that I won't ever use this nifty little module. 😭

**Reply by rchiileea** [reply]
they done it like this as its the nature of the beast, the product and uses for need of a dual tpu are mainly industrial with the end user market probably representing about 1% of sales.

**Reply by jlarnal** [reply]
Yes, that was my point : any motherboard or SBC that would offer a fully-complemented M.2 Key-E interface are industrial products (in the $3K price range, obviously, it's a rugged class of appliances ). And that's not a bad thing, imho. 

But that's where the rubber meet the road : most of these SBCs already have NPUs... So the Coral would be a bit... redundant ? 
Unless the point is to use the energy-efficient Coral rather than the onboard, often power-hungry NPU ? In an industrial application ? Meh ? Too much discrepancies in this line of toughs.

**Comment by linuxdep** [comment]
Hi, it is possible to add the M.2 board to WIFI M.2 connector on HP ProDek 600 G4?

**Reply by rchiileea** [reply]
it will but if its the dual version, you wont make much use of it, due to the nature of the way it needs the bandwidth for dual tpu, your best bet would be if dual is to purchase the m.2 adaptor for it and use it in m.2 port and then use a normal sata ssd for the main drive.

**Reply by linuxdep** [reply]
Thanks, I'm not that deep into the subject, but a TPU is certainly enough for me for now. The two M.2 slots are fitted with SSDs, which I can't and don't want to remove, but if the slot for the WIFI card also works, I think it's better than USB, don't you?

**Comment by noelhibbard** [comment]
I am using the Dual Edge in Frigate with no issues other than my adapter (made for an M.2 WiFi NIC) is only giving me access to one TPU. The key to getting the Dual Edge to work is by getting the right adapter. This $30 adapter will give you access to both TPUs. https://www.makerfabs.com/dual-edge-tpu-adapter.html

**Comment by dataslayermedia** [comment]
The Dual Edge TPU is working well for me with my [Zima Blade](https://shop.zimaboard.com/products/zimablade-single-board-server-for-cyber-native?utm_source=productpage&utm_medium=link&variant=47722451730724) and [Zima Board](https://amzn.to/3IjEL7N)...


<img width="471" alt="image" src="https://github.com/google-coral/edgetpu/assets/731746/a9b9a6e8-9ca3-411d-9b0b-5c3ea677e432">

I'm able to access both TPU's from the [Coral Dual PCIe accelerator](https://amzn.to/49QcIIQ) using the [Maker Fab Dual TPU Low Profile adapter](https://www.makerfabs.com/dual-edge-tpu-adapter.html) on the [Zima Blade](https://shop.zimaboard.com/products/zimablade-single-board-server-for-cyber-native?utm_source=productpage&utm_medium=link&variant=47722451730724) and [Zima Board](https://amzn.to/3IjEL7N). 

Running `lspci -nn | grep 089a` and `ls /dev/apex_*`

<img width="831" alt="image" src="https://github.com/google-coral/edgetpu/assets/731746/93ae8c21-72fd-463c-a311-ea82036e233b">
Operating System: Debian 12

Also works fine with Frigate just needed to pass both apex devices to the docker container.

```
docker run -d \
  --name frigate \
  --restart=unless-stopped \
  --mount type=tmpfs,target=/tmp/cache,tmpfs-size=1000000000 \
  --device /dev/apex_0:/dev/apex_0 \
  --device /dev/apex_1:/dev/apex_1 \
  --shm-size=512m \
  -v /frigate/storage:/media/frigate \
  -v /frigate/config.yml:/config/config.yml \
  -v /etc/localtime:/etc/localtime:ro \
  -p 5000:5000 \
  -p 8555:8555/tcp \
  -p 8555:8555/udp \
  ghcr.io/blakeblackshear/frigate:stable
```

config.yml
```
detectors:
  coral1:
    type: edgetpu
    device: pci:0
  coral2:
    type: edgetpu
    device: pci:1
```

[Full Tutorial Blog Post](https://medium.com/@timothydmoody/how-to-setup-the-coral-dual-edge-tpu-for-frigate-9a7f645dc915)

[YouTube Tutorial](https://www.youtube.com/watch?v=W-LlUe_EXjg)

**Reply by didi767** [reply]
Hi, I have the adapter but still only one TPU works. (it doesn't add the name 089a to each of the TPUs, does it matter?)
output for the commands:
.
sudo lspci:
00:10.0 System peripheral: Global Unichip Corp. Coral Edge TPU
00:11.0 System peripheral: Global Unichip Corp. Coral Edge TPU

user@frigate-242:~$ sudo ls /dev/apex_*
/dev/apex_0  /dev/apex_1

frigate only recognize pci0

detectors:
  coral_pci1:
    type: edgetpu
    device: pci:0
  # coral_pci2:
  #     type: edgetpu
  #     device: pci:1

**Comment by serg987** [comment]
Made a [small adapter](https://github.com/serg987/coral-dual-m2-adapter-pcb) to use 2 dual Coral key E modules with key M ASM1184 board widely available on marketplaces. It may not fit setups with limited space but works if there's around 4 cm above the key M slot. It allows to use 4 TPUs at once, and worked in Frigate till I switched to another TPU.
Looks weird, but works
![](https://github.com/serg987/coral-dual-m2-adapter-pcb/raw/master/media/chatreey_installation_new.jpg)

**Comment by zolakt** [comment]
I have the dual TPU Coral. So far I had it plugged into the m2 slot, but since it has 1 lane, it was detecting only 1 tpu. Now I got [this adapter ](https://www.makerfabs.com/dual-edge-tpu-adapter.html), but whatever I try I can't get this to work in Frigate. I'm running Frigate in a Proxmox privileged LXC. I know it's not the advised way to run, but I had zero issues so far.

Anyway, proxmox host detects the adapter, and the 2 tpus:
```
root@proxmox:~# ls -l /dev/apex*
crw-rw---- 1 root root 120, 0 Nov 18 13:49 /dev/apex_0
crw-rw---- 1 root root 120, 1 Nov 18 13:49 /dev/apex_1

root@proxmox:~# lspci -nn | grep 089a
06:00.0 System peripheral [0880]: Global Unichip Corp. Coral Edge TPU [1ac1:089a]
07:00.0 System peripheral [0880]: Global Unichip Corp. Coral Edge TPU [1ac1:089a]
```

I'm passing both to the LXC with the simple device passthrough:
```
arch: amd64
cores: 16
dev0: /dev/apex_0
dev1: /dev/apex_1
dev2: /dev/dri/card0
dev3: /dev/dri/renderD128
dev4: /dev/nvidia0
dev5: /dev/nvidiactl
dev6: /dev/nvidia-uvm
dev7: /dev/nvidia-uvm-tools
dev8: /dev/nvidia-caps/nvidia-cap1
dev9: /dev/nvidia-caps/nvidia-cap2
features: nesting=1,fuse=1
hostname: media-server
memory: 32768
net0: name=eth0,bridge=vmbr0,hwaddr=BC:24:11:31:1D:42,ip=dhcp,tag=20,type=veth
onboot: 1
ostype: debian
rootfs: local-lvm:vm-2000-disk-0,size=40G
swap: 512
lxc.net.0.flags: up
```

And both tpus are visible in the LXC:
```
root@media-server:~# ls -l /dev/apex*
crw-rw---- 1 root root 120, 0 Nov 18 14:30 /dev/apex_0
crw-rw---- 1 root root 120, 1 Nov 18 14:30 /dev/apex_1

root@media-server:~# udevadm info --name=/dev/apex_0
P: /devices/pci0000:00/0000:00:1c.0/0000:04:00.0/0000:05:03.0/0000:06:00.0/apex/apex_0
M: apex_0
R: 0
J: c120:0
U: apex
D: c 120:0
N: apex_0
L: 0
E: DEVPATH=/devices/pci0000:00/0000:00:1c.0/0000:04:00.0/0000:05:03.0/0000:06:00.0/apex/apex_0
E: DEVNAME=/dev/apex_0
E: MAJOR=120
E: MINOR=0
E: SUBSYSTEM=apex

root@media-server:~# udevadm info --name=/dev/apex_1
P: /devices/pci0000:00/0000:00:1c.0/0000:04:00.0/0000:05:07.0/0000:07:00.0/apex/apex_1
M: apex_1
R: 1
J: c120:1
U: apex
D: c 120:1
N: apex_1
L: 0
E: DEVPATH=/devices/pci0000:00/0000:00:1c.0/0000:04:00.0/0000:05:07.0/0000:07:00.0/apex/apex_1
E: DEVNAME=/dev/apex_1
E: MAJOR=120
E: MINOR=1
E: SUBSYSTEM=apex
```

Then I pass both of them to docker:
```
    devices:
      - /dev/apex_0:/dev/apex_0
      - /dev/apex_1:/dev/apex_1
```

So far so good... But in Frigate config as soon as I try to use `device: pci:1` it crashes with this error and keeps restarting:
```
2025-11-18 15:54:37.187559395  [2025-11-18 15:54:37] frigate.detectors.plugins.edgetpu_tfl ERROR   : No EdgeTPU was detected. If you do not have a Coral device yet, you must configure CPU detectors.
2025-11-18 15:54:37.197176477  Process detector:coral1:
2025-11-18 15:54:37.197186204  Traceback (most recent call last):
2025-11-18 15:54:37.197190719    File "/usr/local/lib/python3.11/dist-packages/tflite_runtime/interpreter.py", line 166, in load_delegate
2025-11-18 15:54:37.197194693      delegate = Delegate(library, options)
2025-11-18 15:54:37.197198919                 ^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-11-18 15:54:37.197203515    File "/usr/local/lib/python3.11/dist-packages/tflite_runtime/interpreter.py", line 104, in __init__
2025-11-18 15:54:37.197250571      raise ValueError(capture.message)
2025-11-18 15:54:37.197254157  ValueError
2025-11-18 15:54:37.615124  2025-11-18 15:54:37.197259428
2025-11-18 15:54:37.197263351  During handling of the above exception, another exception occurred:
2025-11-18 15:54:37.615131  2025-11-18 15:54:37.197266790
2025-11-18 15:54:37.197270275  Traceback (most recent call last):
2025-11-18 15:54:37.197275106    File "/usr/lib/python3.11/multiprocessing/process.py", line 314, in _bootstrap
2025-11-18 15:54:37.197378289      self.run()
2025-11-18 15:54:37.197383341    File "/opt/frigate/frigate/util/process.py", line 41, in run_wrapper
2025-11-18 15:54:37.197387446      return run(*args, **kwargs)
2025-11-18 15:54:37.197390691             ^^^^^^^^^^^^^^^^^^^^
2025-11-18 15:54:37.197394669    File "/usr/lib/python3.11/multiprocessing/process.py", line 108, in run
2025-11-18 15:54:37.197427565      self._target(*self._args, **self._kwargs)
2025-11-18 15:54:37.197431481    File "/opt/frigate/frigate/object_detection/base.py", line 112, in run_detector
2025-11-18 15:54:37.197435827      object_detector = LocalObjectDetector(detector_config=detector_config)
2025-11-18 15:54:37.197439266                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-11-18 15:54:37.197445472    File "/opt/frigate/frigate/object_detection/base.py", line 57, in __init__
2025-11-18 15:54:37.197449551      self.detect_api = create_detector(detector_config)
2025-11-18 15:54:37.197453680                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-11-18 15:54:37.197498628    File "/opt/frigate/frigate/detectors/__init__.py", line 18, in create_detector
2025-11-18 15:54:37.197503941      return api(detector_config)
2025-11-18 15:54:37.197508252             ^^^^^^^^^^^^^^^^^^^^
2025-11-18 15:54:37.197513899    File "/opt/frigate/frigate/detectors/plugins/edgetpu_tfl.py", line 42, in __init__
2025-11-18 15:54:37.197521119      edge_tpu_delegate = load_delegate("libedgetpu.so.1.0", device_config)
2025-11-18 15:54:37.197526856                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-11-18 15:54:37.197532942    File "/usr/local/lib/python3.11/dist-packages/tflite_runtime/interpreter.py", line 168, in load_delegate
2025-11-18 15:54:37.197580081      raise ValueError('Failed to load delegate from {}\n{}'.format(
2025-11-18 15:54:37.197585348  ValueError: Failed to load delegate from libedgetpu.so.1.0
2025-11-18 15:54:37.615141  2025-11-18 15:54:37.197589783
```

The config I'm trying to use is:
```
detectors:
  coral1:
    type: edgetpu
    device: pci:0
  coral2:
    type: edgetpu
    device: pci:1
```

I tried flipping the order, as suggested in one of the posts here... it didn't help.

If I just use one of them with `device: pci` or `device: pci:0` everything works fine. But as soon as `pci:1` comes into play, it crashes, even when I just try to use one of them with `pci:1`. So even this doesnt't work:
```
detectors:
  coral1:
    type: edgetpu
    device: pci:1
```

I'm completely out of ideas. Other than overheating, although I don't think that is the issue. If just `pci:0` works fine, why wouldn't `pci:1` work on it's own? When I just use one tpu the temperature is around 40C.

Does anyone have a suggestion?

**Reply by zolakt** [reply]
Maybe worth noting. I've tried to check the temperatures. The second tpu is showing -90C, which makes no sense. Maybe this is the problem?

```
root@proxmox:~# cat /sys/class/apex/apex_0/temp
39550
root@proxmox:~# cat /sys/class/apex/apex_1/temp
-89700
root@proxmox:~# cat /sys/class/apex/apex_0/hw_temp_warn2
99800
root@proxmox:~# cat /sys/class/apex/apex_1/hw_temp_warn2
99800
root@proxmox:~# cat /sys/class/apex/apex_0/hw_temp_warn2_en
1
root@proxmox:~# cat /sys/class/apex/apex_1/hw_temp_warn2_en
1
```

**Reply by zolakt** [reply]
It looks like this was a physical problem. The adapter is pretty loose in the pcie slot, without anything to hold it. So I 3d printed a bracket and attached the adapter to the case with it. After that both tpus work.

It's even a pretty crappy thin bracket, but it makes enough difference. It will have to do until the aluminum one I ordered arrives. These adapters should really come with brackets included to avoid issues like this. It costs almost nothing.

**Reply by deanpm70** [reply]
Your configuration looks perfect. The devices appear to be present and correct. The error you're seeing suggests that Frigate cannot load the libedgetpu.so.1.0 library so I'm wondering if this is simply a driver issue. I suggest checking for its presence in your container with something like "locate libedgetpu.so.1.0" or "find /usr/lib -name libedgetpu.so.1.0".

My guess is that it's not there in which case you would need to download and install (or build) the driver. Google returns a few useful looking hits for doing this.

**Reply by zolakt** [reply]
It's not the driver. As I've mentioned above, once I fixated the adapter with a brackert to the pc case, everything started to work in Frigate.

However, I'm now getting [this issue](https://github.com/google-coral/edgetpu/issues/345#issuecomment-3556626970) from time to time. Occassionally my whole proxmox host freezes. And just before the freeze I start seeing these errors from the tpu. Never had that before putting this adapter in. And I had the coral working for over a year in the m2 slot, never any problems with it.

Honestly, I just think this adapter is poorly made. Weight is poorly balanced, and it looses connection. Probably fine in a horizontal pc case, but it a vertical one it just wobbles around in the pci slot too much. And/or the chip on the adapter overheats. This should really come with a bracket included, or ideally with some addon that would fill longer pci lanes and keep it fixated.

Either way, I'm trying "the fix" mentioned in the thread I linked. If I get a freeze again, it's back to a single tpu, and the adapter goes to the trash bin

**Reply by FrigateEnthusiast** [reply]
As far as temps go... You find and order from Amazon, small copper heatsinks with sticky thermal pads. Search: "small copper heatsinks".

Place one on each tpu chip, one on the adapter chip. Then temps should NEVER, EVER be an an issue. Of course this will only work if you have adequate airflow across your PCI adapter card. And if you DON'T HAVE THAT, YOU DESERVE PAIN AND SUFFERING, FAIL! lol!

**Reply by zolakt** [reply]
I doubt this is a temperature issue. My coral almost never goes over 45C. I also had this freeze issue when the coral wasn't even plugged into the adapter. I think, the only possibility of overheating is with the adapter chip itself.

I might try heatsinks, although I very much doubt that is the issue. Honestly, I'm tempted just to throw this in the garbage without loosing more time and nerves on it. Pretty pissed off with it... Why isn't the bracket included when It doesn't sit properly in the slot... Why doesn't it come with a heatsink if it's known to have overheating issues... both of those combined don't cost more than a few $...

Btw. do you know what dimensions these heatsinks should be? "small" is a relative term, and I really don't want to take this thing out again just to measure it, then loose half of the day again to get it properly detected.

**Reply by FrigateEnthusiast** [reply]
You said this: "And/or the chip on the adapter overheats.", so I'm responding to that.

As far as "do you know what dimensions these heatsinks should be?", how would I know what your adapter chip size is?

For the Coral EdgeTPU, there is very extensive information on Thermal Management in the datasheet. That's where you'll find your answers.

**Reply by zolakt** [reply]
I figured out what happened. Unfortunately it seems my coral got damaged. The host freezes during/after detection. Sometimes it works a few times, but more often that not, it freezes the host. And it's now doing that even without the pci adapter, when plugged to the m2 port directly. My assumption is that all of these connection problems with the adapter eventually fried something... Sorry I ever bought the damn thing, definetly wouldn't advise it to anyone.

I had a dilemma for a while if I should switch from the coral to the iGPU and openvino, and now I'm forced to make the switch. Time to retire the coral unfortunately.
